internal/query/roles_v2.go

```diff
@@ -0,0 +1,623 @@
+	role.Actions = newActions
+
+	return role, nil
+}
+
+func (e *engine) DeleteRoleV2(ctx context.Context, roleResource types.Resource) error {
+	ctx, span := e.tracer.Start(ctx, "engine.DeleteRoleV2")
+	defer span.End()
+
+	dbCtx, err := e.store.BeginContext(ctx)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return err
+	}
+
+	err = e.store.LockRoleForUpdate(dbCtx, roleResource.ID)
+	if err != nil {
+		sErr := fmt.Errorf("failed to lock role: %s: %w", roleResource.ID, err)
+
+		span.RecordError(sErr)
+		span.SetStatus(codes.Error, sErr.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return err
+	}
+
+	dbRole, err := e.store.GetRoleByID(dbCtx, roleResource.ID)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return err
+	}
+
+	roleOwner, err := e.NewResourceFromID(dbRole.ResourceID)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return err
+	}
+
+	// 1. delete role from permission-api DB
+	if _, err = e.store.DeleteRole(dbCtx, roleResource.ID); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return err
+	}
+
+	// 2. delete role relationships from spice db
+	errs := []error{}
+
+	// 2.a remove all relationships from this role
+
+	delRoleRelationshipReq := &pb.DeleteRelationshipsRequest{
+		RelationshipFilter: &pb.RelationshipFilter{
+			ResourceType:       e.namespaced(e.rbac.RoleResource.Name),
+			OptionalResourceId: roleResource.ID.String(),
+		},
+	}
+
+	if _, err := e.client.DeleteRelationships(ctx, delRoleRelationshipReq); err != nil {
+		errs = append(errs, err)
+	}
+
+	// 2.b remove all relationships to this role from its owner
+
+	ownerRelReq := &pb.DeleteRelationshipsRequest{
+		RelationshipFilter: &pb.RelationshipFilter{
+			ResourceType: e.namespaced(roleOwner.Type),
+			OptionalSubjectFilter: &pb.SubjectFilter{
+				SubjectType:       e.namespaced(e.rbac.RoleResource.Name),
+				OptionalSubjectId: roleResource.ID.String(),
+			},
+		},
+	}
+
+	if _, err := e.client.DeleteRelationships(ctx, ownerRelReq); err != nil {
+		errs = append(errs, err)
+	}
+
+	// 2.c TODO remove all role relationships in role bindings associated with this role
+
+	for _, err := range errs {
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+
+			logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+			return err
+		}
+	}
+
+	if err = e.store.CommitContext(dbCtx); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		// At this point, spicedb changes have already been applied.
+		// Attempting to rollback could result in failures that could result in the same situation.
+		//
+		// TODO: add spicedb rollback logic along with rollback failure scenarios.
```

Comment:
added the following:

> In this particular case, in the absence of spiceDB rollback, the expected
> behavior in this case would be that the role only exists in the permissions-api
> DB and not in spiceDB. This would result in the role being "orphaned"
> (lack of ownership in spiceDB will prevent this role from being accessed or bind
> by any user), and would need to be cleaned up manually.